src/plot_api/plot_api.js

```diff
@@ -293,12 +293,21 @@ Plotly.plot = function(gd, data, layout, config) {
         // in case of traces that were heatmaps or contour maps
         // previously, remove them and their colorbars explicitly
         for (i = 0; i < calcdata.length; i++) {
-            cd = calcdata[i];
-            trace = cd[0].trace;
-            if (trace.visible !== true || !trace._module.colorbar) {
+            trace = calcdata[i][0].trace;
+
+            var isVisible = (trace.visible === true),
                 uid = trace.uid;
-                fullLayout._paper.selectAll('.hm'+uid+',.contour'+uid+',.cb'+uid+',#clip'+uid)
-                    .remove();
+
+            if(!isVisible || !Plots.traceIs(trace, '2dMap')) {
```

Comment:
we have to compare `trace.visible === true`, because `visible` can also be set to `'legendonly'`.

So I thought of using `isVisible` instead of having two `trace.visible === true` checks.